Use single webpack config, /webpack.config.js for rails/webpacker, matching jsbundling-rails (#3240)
rails/webpacker: docs/v6_upgrade.md

This means you have to configure integration with frameworks yourself, but webpack-merge helps with this. See this example for [Vue](https://github.com/rails/webpacker#other-frameworks) and scroll to the bottom for [more examples](#examples-of-v5-to-v6).

## How to upgrade to Webpacker v6 from v5

> 1. Ensure you have a clean working git branch. You will be overwriting all your files and reverting the changes that you don't want.

Reviewer: ```suggestion
1. Ensure you have a clean working git branch. You will be overwriting all your files and reverting the changes that you don't want. It's recommended removing `node_modules` directory and running `rails assets:clobber` before upgrading.
```

> 1. Upgrade the Webpacker Ruby gem and the NPM package

Reviewer: ```suggestion
2. Upgrade the Webpacker Ruby gem and the NPM package
```

Author: Using `1.` everywhere results in less merge conflicts! 

@guillaumebriday, @dhh Do you agree that only `1` is better for a numbered list?

   Note: [Check the releases page to verify the latest version](https://github.com/rails/webpacker/releases), and make sure to install identical version numbers of webpacker gem and `@rails/webpacker` npm package. (Gems use a period and packages use a dot between the main version number and the beta version.)

   Example going to a specific version:

   ```ruby
   # Gemfile
   gem 'webpacker', '6.0.0.rc.7'
   ```

   ```bash
   bundle install
   ```

   ```bash
   yarn add @rails/webpacker@6.0.0-rc.7 --exact
   ```

   ```bash
   bundle exec rails webpacker:install
   ```

1. There is now a single default configuration file of `config/webpack/webpack.config.js`. Previously, the config file was set

>   to `config/webpack/#{NODE_ENV}.js`. In the `config/webpack/` directory, you can either refactor your code in `test.js`, `development.js`, and `production.js` to a single file, `config.webpack.js` or you can replace the contents of `config/webpack/config.webpack.js` to conditionally load the old file based on the NODE_ENV with this snippet:

Reviewer 2: i think `config.webpack.js` should be `webpack.config.js`

Author: I agree! Per the docs: https://webpack.js.org/configuration/, `webpack.config.js` is what's used.

   ```js
   const { env, webpackConfig } = require('@rails/webpacker')
   const { existsSync } = require('fs')
   const { resolve } = require('path')
   
   const envSpecificConfig = () => {
     const path = resolve(__dirname, `${env.nodeEnv}.js`)
     if (existsSync(path)) {
       console.log(`Loading ENV specific webpack configuration file ${path}`)
       return require(path)
     } else {
       return webpackConfig
     }
   }
   
   module.exports = envSpecificConfig()
   ```

> 1. Review the new default's changes to `webpacker.yml`. Consider each suggested change carefully, especially the change to have your `source_entry_path` be at the top level of your `source_path`. 

Reviewer: ```suggestion
3. Review the new default's changes to `webpacker.yml`. Consider each suggested change carefully, especially the change to have your `source_entry_path` be at the top level of your `source_path`. 
```

   The v5 default used `packs` for `source_entry_path`:
   ```yml
   source_path: app/javascript
   source_entry_path: packs
   ```
   The v6 default uses the top level directory.
   ```yml
   source_path: app/javascript
   source_entry_path: /
   ```
   If you prefer this configuratiom, then you will move your `app/javascript/packs/*` (including `application.js`) to `app/javascript/` and update the configuration file. 
  
   Note, moving your files is optional, as you can stil keep your entries in a separate directory, called something like `packs`, or `entries`. This directory is defined with the `source_path`.
  

> 1. **Ensure no nested directories in your `source_entry_path`.** Check if you had any entry point files in child directories of your `source_entry_path`. Files for entry points in child directories are not supported by rails/webpacker v6. Move those files to the top level, adjusting any imports in those files.

Reviewer: ```suggestion
4. **Ensure no nested directories in your `source_entry_path`.** Check if you had any entry point files in child directories of your `source_entry_path`. Files for entry points in child directories are not supported by rails/webpacker v6. Move those files to the top level, adjusting any imports in those files.
```

  
   The new v6 configuration does not allow nesting, so as to allow placing the entry points at in the root directory of JavaScript. You can find this change [here](https://github.com/rails/webpacker/commit/5de0fbc1e16d3db0c93202fb39f5b4d80582c682#diff-7af8667a3e36201db57c02b68dd8651883d7bfc00dc9653661be11cd31feeccdL19).

> 1. Update `webpack-dev-server` to the current version, greater than 4.2, updating `package.json`.

Reviewer: ```suggestion
5. Update `webpack-dev-server` to the current version, greater than 4.2, updating `package.json`.
```

> 1. Update API usage of the view helpers by changing `javascript_packs_with_chunks_tag` and `stylesheet_packs_with_chunks_tag` to `javascript_pack_tag` and `stylesheet_pack_tag`. Ensure that your layouts and views will only have **at most one call** to `javascript_pack_tag` and **at most one call** to `stylesheet_pack_tag`. You can now pass multiple bundles to these view helper methods. If you fail to changes this, you may experience performance issues, and other bugs related to multiple copies of React, like [issue 2932](https://github.com/rails/webpacker/issues/2932).  If you expose jquery globally with `expose-loader,` by using `import $ from "expose-loader?exposes=$,jQuery!jquery"` in your `app/javascript/application.js`, pass the option `defer: false` to your `javascript_pack_tag`.

Reviewer: ```suggestion
6. Update API usage of the view helpers by changing `javascript_packs_with_chunks_tag` and `stylesheet_packs_with_chunks_tag` to `javascript_pack_tag` and `stylesheet_pack_tag`. Ensure that your layouts and views will only have **at most one call** to `javascript_pack_tag` and **at most one call** to `stylesheet_pack_tag`. You can now pass multiple bundles to these view helper methods. If you fail to changes this, you may experience performance issues, and other bugs related to multiple copies of React, like [issue 2932](https://github.com/rails/webpacker/issues/2932).  If you expose jquery globally with `expose-loader,` by using `import $ from "expose-loader?exposes=$,jQuery!jquery"` in your `app/javascript/application.js`, pass the option `defer: false` to your `javascript_pack_tag`.
```

> 1. If you are using any integrations like `css`, `postcss`, `React` or `TypeScript`. Please see https://github.com/rails/webpacker#integrations section on how they work in v6.

Reviewer: ```suggestion
7. If you are using any integrations like `css`, `postcss`, `React` or `TypeScript`. Please see https://github.com/rails/webpacker#integrations section on how they work in v6.
```

> 1. Import `environment` changed to `webpackConfig`. For example, the new code looks like:

Reviewer: ```suggestion
8. Import `environment` changed to `webpackConfig`. For example, the new code looks like:
```

   ```js
   // config/webpack/webpack.config.js
   const { webpackConfig, merge } = require('@rails/webpacker')
   const customConfig = require('./custom')

   module.exports = merge(webpackConfig, customConfig)
   ```

> 1. Copy over custom browserlist config from `.browserslistrc` if it exists into the `"browserslist"` key in `package.json` and remove `.browserslistrc`.

Reviewer: ```suggestion
9. Copy over custom browserlist config from `.browserslistrc` if it exists into the `"browserslist"` key in `package.json` and remove `.browserslistrc`.
```

> 1. Remove `babel.config.js` if you never changed it. Configure your `package.json` to use the default:

Reviewer: ```suggestion
10. Remove `babel.config.js` if you never changed it. Configure your `package.json` to use the default:
```

    ```json
    "babel": {
      "presets": [
        "./node_modules/@rails/webpacker/package/babel/preset.js"
      ]
    }
    ```

> 1. `extensions` was removed from the `webpacker.yml` file. Move custom extensions to your configuration by merging an object like this. For more details, see docs for [Webpack Configuration](https://github.com/rails/webpacker/blob/master/README.md#webpack-configuration)

Reviewer: ```suggestion
11. `extensions` was removed from the `webpacker.yml` file. Move custom extensions to your configuration by merging an object like this. For more details, see docs for [Webpack Configuration](https://github.com/rails/webpacker/blob/master/README.md#webpack-configuration)
```

    ```js
    {
      resolve: {
        extensions: ['.ts', '.tsx', '.vue', '.css']
      }
    }
    ```

> 1. Some dependencies were removed in [PR 3056](https://github.com/rails/webpacker/pull/3056). If you see the error: `Error: Cannot find module 'babel-plugin-macros'`, or similar, then you need to `yarn add <dependency>` where <dependency> might include: `babel-plugin-macros`, `case-sensitive-paths-webpack-plugin`, `core-js`, `regenerator-runtime`. Or you might want to remove your dependency on those.

Reviewer: ```suggestion
13. Some dependencies were removed in [PR 3056](https://github.com/rails/webpacker/pull/3056). If you see the error: `Error: Cannot find module 'babel-plugin-macros'`, or similar, then you need to `yarn add <dependency>` where <dependency> might include: `babel-plugin-macros`, `case-sensitive-paths-webpack-plugin`, `core-js`, `regenerator-runtime`. Or you might want to remove your dependency on those.
```

> 1. If `bin/yarn` did not exist, it was added.

Reviewer: ```suggestion
14. If `bin/yarn` does not exist, create an executable [yarn](https://github.com/rails/webpacker/blob/master/lib/install/bin/yarn) file in your `/bin` directory.
```

btw this step is

> 1. Remove overlapping dependencies from your `package.json` and rails/webpacker's `package.json`. For example, don't include `webpack` directly as that's a dependency of rails/webpacker.

Reviewer: ```suggestion
15. Remove overlapping dependencies from your `package.json` and rails/webpacker's `package.json`. For example, don't include `webpack` directly as that's a dependency of rails/webpacker.
```

> 1. Make sure that you can run `bin/webpack` without errors.

Reviewer: ```suggestion
16. Make sure that you can run `bin/webpack` without errors.
```

> 1. Try running `RAILS_ENV=production bin/rails assets:precompile`. If all goes well, don't forget to clean the generated assets with `bin/rails assets:clobber`.

Reviewer: ```suggestion
17. Test your production build locally by running `RAILS_ENV=production bin/rails assets:precompile`. If all goes well, don't forget to clean the generated assets with `bin/rails assets:clobber`.
```

> 1. Run `yarn add webpack-dev-server` if those are not already in your dev dependencies. Make sure you're using v4+.

Reviewer: ```suggestion
18. Run `yarn add -D webpack-dev-server` if those are not already in your dev dependencies. Make sure you're using v4+.
```

> 1. Try your app!

Reviewer: ```suggestion
19. Try your app!
```
